docs/RPCS/local_intent_resolution.md

```diff
@@ -0,0 +1,59 @@
+
+An easy way out is adding a sequence number to the `Transaction` object whose single purpose is to be incremented by the client before sending another request in a `Txn`: The gateway aborts the `Transaction` if the sequence number does not match; additionally, a Transaction can only start with sequence number one. The returned error indicates an abort of the `Txn` with immediate retry.
+
+This provides a simple method for clients to stick to one coordinator, and gives us an authoritative list of intents. Any intent not in the list is considered aborted; that way ill-behaved clients may lose writes, but otherwise we would not be able to gc `Txn` records ever.
+
+## Gateway changes
+
+The logic for resolving intents on `EndTransaction` is changed: Instead of trying to resolve the intents, their keys are bundled up and sent along with the `EndTransaction` call. Once that call has returned, the coordinator's work for that `Txn` is done.
+
+Some updates are necessary for tracking the sequence number:
+
+* a `Txn` is only initiated with the first mutating command, but this may not have sequence number zero.
+* heartbeating a `Transaction` which just got committed could recreate the freshly-gc'ed record; some care needs to be taken that only the first heartbeat can create a record (all others should `CPut`), and that the heartbeat goroutine ends before the gateway passes on the `EndTransaction` call. This isn't a correctness issue, though.
+
+## Range changes
+
+The logic for `EndTransaction` is updated to reflect the following enhancements:
+
+* the intent list is split into range-local and non-local intents.
+* all local intents are atomically added to the `EndTransaction` batch
+* the transaction record is updated so that its authoritative list of intents contains only the non-local intents; if there are no non-local intents, the record can be gc'ed immediately.
+* after the batch commits and only on the leading replica, the unresolved intents are resolved in a `Batch` asynchronously. Once resolved, the `Txn` record should be updated (as with any intent resolution; the functionality can be factored out into a single location). If the `Txn` is without intents after the update, it can be removed.
+* The GC queue (or a new queue) should groom the `Transaction` records of each `Range` and re-trigger the resolution process described above for all records which have open intents (which will remove them when done). This should only ever have to do actual work in the event of node crashes, etc.
+
+All in all, this should be fairly straightforward and reduces some complexity and shared responsibility regarding intent resolution.
+
+# Drawbacks
+
+Clients will have their `Txn` aborted if switching gateways (likely non-issue).
```

Comment:
I'm slightly concerned about this. At Google, mapreduces would originally have to restart if their (single) master task failed. This was enough of a problem that they needed to implement master failover (eventually - it was tolerated for a long time). But since allowing transactions to cross coordinators seems expensive (either the existing issues with lazy intent resolution or storing those intents somewhere), I'm fine proceeding with this.